core/src/socket.rs

```diff
@@ -0,0 +1,217 @@
+        //       or SocketAction::Failed when connection failed.
+        backend.connect_socket(host, port, handle, receiver, self.sender.clone());
+
+        if let Some(existing_handle) = target.set_handle(handle) {
+            // As written in the AS3 docs, we are supposed to close the existing connection,
+            // when a new one is created.
+            self.close(existing_handle)
+        }
+    }
+
+    pub fn is_connected(&self, handle: SocketHandle) -> bool {
+        matches!(self.sockets.get(handle), Some(Socket { .. }))
+    }
+
+    pub fn send(&mut self, handle: SocketHandle, data: Vec<u8>) {
+        if let Some(Socket { send_buffer, .. }) = self.sockets.get_mut(handle) {
+            send_buffer.push_back(data);
+        }
+    }
+
+    pub fn close(&mut self, handle: SocketHandle) {
+        if let Some(Socket { sender, .. }) = self.sockets.remove(handle) {
+            drop(sender); // NOTE: By dropping the sender, the reading task will close automatically.
+        }
+    }
+
+    pub fn update_sockets(context: &mut UpdateContext<'_, 'gc>) {
+        let mut activation = Activation::from_nothing(context.reborrow());
+
+        let mut actions = vec![];
+
+        while let Ok(action) = activation.context.sockets.receiver.try_recv() {
+            actions.push(action)
+        }
+
+        for action in actions {
+            match action {
+                SocketAction::Connect(handle, success) => {
+                    if success {
+                        let target = activation
+                            .context
+                            .sockets
+                            .sockets
+                            .get(handle)
+                            .expect("only valid handles in SocketAction")
+                            .target;
+
+                        let connect_evt =
+                            EventObject::bare_default_event(&mut activation.context, "connect");
+                        Avm2::dispatch_event(&mut activation.context, connect_evt, target.into());
+                    } else {
+                        let target = activation
+                            .context
+                            .sockets
+                            .sockets
+                            .get(handle)
+                            .expect("only valid handles in SocketAction")
+                            .target;
+
+                        let io_error_evt = activation
+                            .avm2()
+                            .classes()
+                            .ioerrorevent
+                            .construct(
+                                &mut activation,
+                                &[
+                                    "ioError".into(),
+                                    false.into(),
+                                    false.into(),
+                                    "TODO".into(),
+                                    0.into(),
```

Comment:
Not sure what message or id to use here.

Reply:
For now this can use `Error #2031: Socket Error.`